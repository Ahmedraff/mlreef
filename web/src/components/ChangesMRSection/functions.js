import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
