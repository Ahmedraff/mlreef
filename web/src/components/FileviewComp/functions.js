import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
