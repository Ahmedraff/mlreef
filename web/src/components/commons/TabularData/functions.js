import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
