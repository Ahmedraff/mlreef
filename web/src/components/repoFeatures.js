import React, { Component } from "react";
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import arrow_down_blue_01 from "./../images/arrow_down_blue_01.svg";
import plus_01 from "./../images/plus_01.svg";
import { Link } from "react-router-dom";
import * as branchesActions from "./../actions/branchesActions";

class RepoFeatures extends Component {
  constructor(props){
    super(props);
    this.state = {
      isOpen: false,
      plusOpen: false,
      branchSelected: decodeURIComponent(this.props.branch),
      projectId: this.props.projects.selectedProject.id,
      branches: []
    };

    this.props.actions.getBranchesList(this.state.projectId);
  }

  branchRef = React.createRef();
  plusRef = React.createRef();

  handleBlur = e => {
    this.handleBranch();
  };

  plusDropdownBlur = e => {
    this.plusDropdown();
  };

  handleBranch = e => {
    if (!this.state.isOpen) {
      document.addEventListener("click", this.handleBlur, false);
    } else {
      document.removeEventListener("click", this.handleBlur, false);
    }

    this.setState(prevState => ({
      isOpen: !prevState.isOpen
    }));
  };

  plusDropdown = e => {
    if (!this.state.plusOpen) {
      document.addEventListener("click", this.plusDropdownBlur, false);
    } else {
      document.removeEventListener("click", this.plusDropdownBlur, false);
    }

    this.setState(prevState => ({
      plusOpen: !prevState.plusOpen
    }));
  };

  componentWillUnmount() {
    this.setState = (state, callback) => {
      return;
    };
  }

  static getDerivedStateFromProps = (nextProps, prevState) => {
    const newState = {...prevState};
    newState.branches = nextProps.branches;
    
    newState.branchSelected = nextProps.branch !== prevState.branchSelected 
      ? nextProps.branch
      : newState.branchSelected;
   
    return newState;
  }

  handleClick = (e) => {
    this.props.actions.loadFiles(
      null,
      encodeURIComponent(e.currentTarget.id),
      this.state.projectId,
      true
    );
  }

  render = () => (
      <>
        <div id="repo-features">
          <div>
            <div className="reference" ref={this.branchRef}>
              <button
                className="white-button"
                onClick={this.handleBranch}
              >
                <span>{this.state.branchSelected}</span>
                <img id="leftfeature-image" src={arrow_down_blue_01} alt="" />
              </button>
              {this.state.isOpen &&
                <div className="select-branch">
                  <div
                    style={{ margin: "0 50px", fontSize: "14px", padding: "0 40px" }}>
                    <p>Switch Branches</p>
                  </div>
                  <hr />
                  <div className="search-branch">
                    <input
                      autoFocus={true}
                      type="text"
                      placeholder="Search branches or tags"
                    />
                    <div className="branches">
                      <ul>
                        <li className="branch-header">Branches</li>
                        {this.state.branches.filter(branch =>
                          !branch.name.startsWith("data-pipeline/") &&
                          !branch.name.startsWith("experiment/")
                        ).map((branch) => {
                          let encoded = encodeURIComponent(branch.name);
                          return (
                            <li key={encoded}>
                              <Link id={branch.name} to={`/my-projects/${this.state.projectId}/${encoded}`} onClick={this.handleClick}><p>{branch.name}</p></Link>
                            </li>
                          )
                        })}
                      </ul>
                    </div>
                  </div>
                </div>
              }
            </div>
            <div className="reference" ref={this.plusRef}>
              <button className="white-button" style={{ position: "relative" }} onClick={this.plusDropdown}>
                <img id="plus" src={plus_01} alt="" />
                <img id="leftfeature-image" src={arrow_down_blue_01} alt="" />
              </button>
              {this.state.plusOpen && <div className="plus-dropdown">
                <ul className="plus-list">
                  <li>This directory</li>
                  <li className="plus-option"><Link to="/">New file</Link></li>
                  <li className="plus-option"><Link to="/">Upload file</Link></li>
                  <li className="plus-option"><Link to="/">New directory</Link></li>
                  <hr />
                  <li>This repository</li>
                  <li className="plus-option"><Link to="/">New branch</Link></li>
                  <li className="plus-option"><Link to="/">New tag</Link></li>
                </ul>
              </div>}
            </div>
            <button className="blue-button">
              <Link to={`/my-projects/${this.state.projectId}/visualizations`}><p>Data Visualisation</p></Link>
            </button>

            <button className="blue-button">
              <Link to={`/my-projects/${this.state.projectId}/pipe-line`}><p>Data Pipeline</p></Link>
            </button>
          </div>
          <div>
            <button className="white-button">
              History
                </button>

            <button className="white-button">
              Web IDE
                </button>

            <button className="white-button">
              <img id="leftfeature-image" src={arrow_down_blue_01} alt="" />
            </button>
          </div>
        </div>
      </>
  );
}

function mapStateToProps(state) {
  return {
    projects: state.projects,
    branches: state.branches
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(branchesActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(RepoFeatures);
