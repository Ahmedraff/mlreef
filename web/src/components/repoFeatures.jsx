import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import {
  string,
  shape,
  number,
  func,
  arrayOf,
} from 'prop-types';
import arrowDown from '../images/arrow_down_blue_01.svg';
import plus01 from '../images/plus_01.svg';
import * as branchesActions from '../actions/branchesActions';
import * as fileActions from '../actions/fileActions';

export class RepoFeatures extends Component {
  branchRef = React.createRef();

  plusRef = React.createRef();

  constructor(props) {
    super(props);
    const {
      branch,
      projects: { selectedProject: { id } },
      actions,
    } = this.props;

    this.state = {
      isOpen: false,
      plusOpen: false,
      branchSelected: decodeURIComponent(branch),
      projectId: id,
      branches: [],
    };
    if(actions){
      actions.getBranchesList(id);
    }
  }

  static getDerivedStateFromProps = (nextProps, prevState) => {
    const newState = { ...prevState };
    newState.branches = nextProps.branches;

    newState.branchSelected = nextProps.branch !== prevState.branchSelected
      ? nextProps.branch
      : newState.branchSelected;

    return newState;
  }

  componentWillUnmount() {
    this.setState = (state) => (state);
  }

  handleBlur = () => {
    this.handleBranch();
  };

  plusDropdownBlur = () => {
    this.plusDropdown();
  };

  handleBranch = () => {
    const { isOpen } = this.state;
    if (!isOpen) {
      document.addEventListener('click', this.handleBlur, false);
    } else {
      document.removeEventListener('click', this.handleBlur, false);
    }

    this.setState((prevState) => ({
      isOpen: !prevState.isOpen,
    }));
  };

  plusDropdown = () => {
    const { plusOpen } = this.state;
    if (!plusOpen) {
      document.addEventListener('click', this.plusDropdownBlur, false);
    } else {
      document.removeEventListener('click', this.plusDropdownBlur, false);
    }

    this.setState((prevState) => ({
      plusOpen: !prevState.plusOpen,
    }));
  };

  handleClick = (e) => {
    const { projectId } = this.state;
    const { updateLastCommit, actions: { loadFiles } } = this.props;
    updateLastCommit(e.currentTarget.id);
    loadFiles(
      null,
      encodeURIComponent(e.currentTarget.id),
      projectId,
      true,
    );
  }

  render() {
    const {
      projectId,
      isOpen,
      plusOpen,
      branchSelected,
      branches,
    } = this.state;
    const { branch, path } = this.props;
    return (
      <>
        {branch !== null ? <div id="repo-features">
          <div>
            <div className="reference" ref={this.branchRef}>
              <button
                type="button"
                id="branch-dropdown"
                className="white-button"
                onClick={this.handleBranch}
              >
                <span>{decodeURIComponent(branchSelected)}</span>
                <img id="leftfeature-image" src={arrowDown} alt="" />
              </button>
              {isOpen && (
                <div id="branches-list" className="select-branch">
                  <div
                    style={{ margin: '0 50px', fontSize: '14px', padding: '0 40px' }}
                  >
                    <p>Switch Branches</p>
                  </div>
                  <hr />
                  <div className="search-branch">
                    <input
                      type="text"
                      placeholder="Search branches or tags"
                    />
                    <div className="branches">
                      <ul>
                        <li className="branch-header">Branches</li>
                        {branches && branches.filter((branch) => !branch.name.startsWith('data-pipeline/')
                          && !branch.name.startsWith('experiment/')).map((branch) => {
                          const encoded = encodeURIComponent(branch.name);
                          return (
                            <li key={encoded}>
                              <Link id={branch.name} to={`/my-projects/${projectId}/${encoded}`} onClick={this.handleClick}><p>{branch.name}</p></Link>
                            </li>
                          );
                        })}
                      </ul>
                    </div>
                  </div>
                </div>
              )}
            </div>
            <div className="reference" ref={this.plusRef}>
              <button
                type="button"
                className="white-button"
                style={{ position: 'relative' }}
                onClick={this.plusDropdown}
              >
                <img id="plus" src={plus01} alt="" />
                <img id="leftfeature-image" src={arrowDown} alt="" />
              </button>
              {plusOpen && (
              <div className="plus-dropdown">
                <ul className="plus-list">
                  <li>This directory</li>
                  <li className="plus-option"><Link to="/">Upload file</Link></li>
                  <li className="plus-option"><Link to="/">New directory</Link></li>
                  <hr />
                  <li>This repository</li>
                  <li className="plus-option"><Link to={`/my-projects/${projectId}/new-branch`}>New branch</Link></li>
                </ul>
              </div>
              )}
            </div>
            <button
              type="button"
              className="blue-button"
            >
              <Link to={`/my-projects/${projectId}/empty-data-visualization`}><p>Data Visualisation</p></Link>
            </button>
            <button
              type="button"
              className="blue-button"
            >
              <Link to={`/my-projects/${projectId}/pipe-line`}><p>Data Pipeline</p></Link>
            </button>
          </div>
          <div>
            <button type="button" className="white-button">
              <Link to={`/my-projects/${projectId}/${branch}/commits/${path}`}>History</Link>
            </button>
          </div>
        </div> : null}
      </>
    );
  }
}

RepoFeatures.propTypes = {
  branch: string.isRequired,
  path: string.isRequired,
  projects: shape({
    selectedProject: shape({
      id: number.isRequired,
    }),
  }).isRequired,
  updateLastCommit: func.isRequired,
  branches: arrayOf(
    shape({
      name: string.isRequired,
    }),
  ).isRequired,
  actions: shape({
    loadFiles: func.isRequired,
    getBranchesList: func.isRequired,
  }).isRequired,
};

function mapStateToProps(state) {
  return {
    projects: state.projects,
    branches: state.branches,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({ ...branchesActions, ...fileActions }, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(RepoFeatures);
