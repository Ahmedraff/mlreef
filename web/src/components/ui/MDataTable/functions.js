const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
