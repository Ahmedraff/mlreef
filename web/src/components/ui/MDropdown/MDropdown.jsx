import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './MDropdown.scss';

const MDropdown = (props) => {
  const {
    className,
    label,
    align,
    listTitle,
    component,
    onClickClose,
    buttonClasses,
    items,
  } = props;

  const [isShown, setIsShown] = useState(false);

  const toggleShow = () => {
    setIsShown(!isShown);
  };

  const close = () => setIsShown(false);

  const handleContainerClick = () => onClickClose && close();

  return (
    <div className={`m-dropdown ${isShown ? 'show' : ''} ${className}`}>
      <div className="m-dropdown-button">
        <button
          type="button"
          className={`${buttonClasses} ${isShown ? 'active' : ''}`}
          onClick={toggleShow}
        >
          {label}
          <i className="fa fa-chevron-down ml-2 my-auto" />
        </button>
      </div>
      <div
        onClick={handleContainerClick}
        className={`m-dropdown-list-container border-rounded ${align} mt-1`}
      >
        {component || (
          <>
            {listTitle && (
              <div className="m-dropdown-list-title">
                {listTitle}
              </div>
            )}
            <ul className="m-dropdown-list p-3">
              {items.map((item, i) => (
                <li key={`dp-item-${i}`} className="m-dropdown-list-item">
                  {item.content}
                </li>
              ))}
            </ul>
          </>
        )}
      </div>
    </div>
  );
};

MDropdown.defaultProps = {
  className: '',
  buttonClasses: 'btn btn-basic-dark px-2',
  align: '',
  listTitle: '',
  label: '',
  onClickClose: true,
  items: [],
  component: null,
};

MDropdown.propTypes = {
  className: PropTypes.string,
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  align: PropTypes.string,
  listTitle: PropTypes.string,
  component: PropTypes.node,
  onClickClose: PropTypes.bool,
  buttonClasses: PropTypes.string,
  items: PropTypes.arrayOf(PropTypes.shape({})),
};

export default MDropdown;
