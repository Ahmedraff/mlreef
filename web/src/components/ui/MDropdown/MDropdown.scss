$m-dropdown-min-width: 300px;

.m-dropdown {
  position: relative;
  width: fit-content;

  &-list {
    list-style: none;
    padding: 0;
    margin-top: 0.25rem;
    margin-bottom: 0;

    &-item {
      padding: 0.5rem;
      cursor: pointer;

      &:hover {
        background-color: var(--light);
      }
    }

    &-title {
      border-bottom: solid 1px var(--secondary);
      padding: 1rem;
      text-align: center;
    }

    &-container {
      min-width: $m-dropdown-min-width;
      background-color: var(--white);
      position: absolute;
      display: none;
      z-index: 2;
      border: none;
      box-shadow: 0 2px 4px #0000008a;

      &.left {
        left: 0;

        & > * {
          left: 0;
        }
      }

      &.right {
        right: 0;

        & > * {
          right: 0;
        }
      }
    }
  }

  &.show {
    .m-dropdown {
      &-button {
        & .fa {
          transform: rotateZ(-180deg);
        }
      }

      &-list {
        &-container {
          display: block;
        }
      }
    }
  }
}
