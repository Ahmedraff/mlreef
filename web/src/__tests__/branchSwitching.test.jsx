import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import MDropdown from 'components/ui/MDropdown';
import { MemoryRouter } from 'react-router-dom';
import { RepoFeatures } from '../components/repoFeatures';
import { projectsArrayMock, branchesMock } from '../testData';
import { storeFactory } from '../functions/testUtils';

const setup = () => shallow(
  <RepoFeatures
    projects={projectsArrayMock.projects}
    branch="master"
    branches={branchesMock}
    path=""
    projectId={projectsArrayMock.projects.selectedProject.gid}
    searchableType={projectsArrayMock.projects.selectedProject.searchableType}
    updateLastCommit={() => {}}
  />,
);

describe('There should render', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });
  it('renders without crashing given the required props', () => {
    const buttonsContainer = wrapper.find('#repo-features');
    expect(buttonsContainer).toHaveLength(1);
  });
});

describe('Dropdown appears on button click', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });
  it('renders without crashing given the required props', () => {
    wrapper
      .find(MDropdown)
      .first()
      .dive()
      .find('.m-dropdown-button button')
      .simulate('click');

    expect(
      wrapper
        .find(MDropdown)
        .first()
        .dive()
        .find('.m-dropdown-list-container'),
    ).toHaveLength(1);
  });
});

describe('The branches list should be displayed when dropdown button is clicked', () => {
  it('dropdown of branches appear', () => {
    const store = storeFactory({
      projects: projectsArrayMock.projects,
      branches: branchesMock,
      user: {
        membership: 1,
        username: 'mlreef',
      },
    });

    const component = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter key="rerere">
            <RepoFeatures
              store={store}
              projects={projectsArrayMock.projects}
              branch="master"
              branches={branchesMock}
              path=""
              projectId={projectsArrayMock.projects.selectedProject.gid}
              searchableType={projectsArrayMock.projects.selectedProject.searchableType}
              updateLastCommit={() => {}}
            />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
