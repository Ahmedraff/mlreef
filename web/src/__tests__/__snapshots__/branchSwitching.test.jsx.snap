// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The branches list should be displayed when dropdown button is clicked dropdown of branches appear 1`] = `
<div
  id="repo-features"
>
  <div
    className="m-dropdown  mr-2 mt-3"
  >
    <div
      className="m-dropdown-button"
    >
      <button
        className="btn btn-basic-dark px-2 "
        onClick={[Function]}
        type="button"
      >
        Master
        <i
          className="fa fa-chevron-down ml-2 my-auto"
        />
      </button>
    </div>
    <div
      className="m-dropdown-list-container border-rounded  mt-1"
      onClick={[Function]}
    >
      <div
        className="select-branch"
        id="branches-list"
      >
        <div
          style={
            Object {
              "fontSize": "14px",
              "margin": "0 50px",
              "padding": "0 40px",
            }
          }
        >
          <p>
            Switch Branches
          </p>
        </div>
        <hr />
        <div
          className="search-branch"
        >
          <input
            placeholder="Search branches or tags"
            type="text"
          />
          <div
            className="branches"
          >
            <ul>
              <li
                className="branch-header"
              >
                Branches
              </li>
              <li>
                <a
                  href="/my-projects/12395599/data-instance%2F664e1c60"
                  id="data-instance/664e1c60"
                  onClick={[Function]}
                >
                  <p>
                    data-instance/664e1c60
                  </p>
                </a>
              </li>
              <li>
                <a
                  href="/my-projects/12395599/data-instance%2F9ae9c8c0"
                  id="data-instance/9ae9c8c0"
                  onClick={[Function]}
                >
                  <p>
                    data-instance/9ae9c8c0
                  </p>
                </a>
              </li>
              <li>
                <a
                  href="/my-projects/12395599/master"
                  id="master"
                  onClick={[Function]}
                >
                  <p>
                    master
                  </p>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="m-dropdown  mr-2 mt-3"
  >
    <div
      className="m-dropdown-button"
    >
      <button
        className="btn btn-basic-dark px-2 "
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa fa-plus"
        />
        <i
          className="fa fa-chevron-down ml-2 my-auto"
        />
      </button>
    </div>
    <div
      className="m-dropdown-list-container border-rounded  mt-1"
      onClick={[Function]}
    >
      <div
        className="plus-dropdown"
      >
        <ul
          className="plus-list"
        >
          <li>
            This directory
          </li>
          <hr />
          <li>
            This repository
          </li>
          <li
            className="plus-option"
          >
            <a
              href="/my-projects/12395599/new-branch"
              onClick={[Function]}
            >
              New branch
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <a
    className="btn btn-dark px-3 mr-2 mt-3"
    href="/my-projects/12395599/pipe-line"
    onClick={[Function]}
  >
    Data Pipeline
  </a>
  <a
    className="btn btn-dark px-3 mr-auto mt-3"
    href="/my-projects/12395599/empty-data-visualization"
    onClick={[Function]}
  >
    Data Visualisation
  </a>
  <a
    className="btn btn-outline-dark mt-3"
    href="/my-projects/12395599/master/commits/"
    onClick={[Function]}
  >
    History
  </a>
</div>
`;
